<launch>
<!-- This script launches both realsense cameras as well as the microstrain imu -->

  <arg name="serial_no_camera"    default="828112073279"/>   <!-- Note: This is the WFI camera -->
  <arg name="camera"             default="camera"/>  
  <arg name="tf_prefix_camera"   default="$(arg camera)"/>
  <arg name="initial_reset"       default="true"/>
  <arg name="serial_no"           default=""/>
  <arg name="json_file_path"      default=""/>
  <arg name="manager"             default="realsense2_camera_manager"/>

  <arg name="fisheye_width"       default="640"/>
  <arg name="fisheye_height"      default="480"/>
  <arg name="enable_fisheye"      default="false"/>

  <arg name="depth_width"         default="640"/>
  <arg name="depth_height"        default="480"/>
  <arg name="enable_depth"        default="true"/>

  <arg name="infra_width"         default="640"/>
  <arg name="infra_height"        default="480"/>
  <arg name="enable_infra1"       default="true"/>
  <arg name="enable_infra2"       default="true"/>

  <arg name="color_width"         default="640"/>
  <arg name="color_height"        default="480"/>
  <arg name="enable_color"        default="true"/>

  <arg name="fisheye_fps"         default="30"/>
  <arg name="depth_fps"           default="5"/>
  <arg name="infra_fps"           default="5"/>
  <arg name="color_fps"           default="5"/>
  <arg name="gyro_fps"            default="400"/>
  <arg name="accel_fps"           default="250"/>
  <arg name="enable_gyro"         default="false"/>
  <arg name="enable_accel"        default="false"/>

  <arg name="enable_pointcloud"   default="false"/>
  <arg name="enable_sync"         default="true"/>
  <arg name="align_depth"         default="true"/>
  <arg name="filters"             default=""/>

  <!-- Arguments for remapping all device namespaces -->
  <arg name="rgb"                             default="color" />
  <arg name="ir"                              default="infra1" />
  <arg name="depth"                           default="depth" />
  <arg name="depth_registered_pub"            default="depth_registered" />
  <arg name="depth_registered"                default="depth_registered" unless="$(arg align_depth)" />
  <arg name="depth_registered"                default="aligned_depth_to_color" if="$(arg align_depth)" />
  <arg name="depth_registered_filtered"       default="$(arg depth_registered)" />
  <arg name="projector"                       default="projector" />

  <!-- Disable bond topics by default -->
  <arg name="bond"                            default="false" />
  <arg name="respawn"                         default="$(arg bond)" />

  <!-- Processing Modules -->
  <arg name="rgb_processing"                  default="true"/>
  <arg name="debayer_processing"              default="false" />
  <arg name="ir_processing"                   default="false"/>
  <arg name="depth_processing"                default="false"/>
  <arg name="depth_registered_processing"     default="true"/>
  <arg name="disparity_processing"            default="false"/>
  <arg name="disparity_registered_processing" default="false"/>
  <arg name="hw_registered_processing"        default="$(arg align_depth)" />
  <arg name="sw_registered_processing"        default="true" unless="$(arg align_depth)" />
  <arg name="sw_registered_processing"        default="false" if="$(arg align_depth)" />

  <!-- DEPTH CAMERA -->
  <group ns="$(arg camera)">

    <include file="$(find realsense2_camera)/launch/includes/nodelet_wfi.launch.xml">
      <arg name="serial_no"                value="$(arg serial_no_camera)"/>
      <arg name="tf_prefix"                value="$(arg tf_prefix_camera)"/>
      <arg name="initial_reset"            value="$(arg initial_reset)"/>
      <arg name="manager"                  value="$(arg manager)"/>
      <arg name="json_file_path"           value="$(arg json_file_path)"/>

      <arg name="enable_pointcloud"        value="$(arg enable_pointcloud)"/>
      <arg name="enable_sync"              value="$(arg enable_sync)"/>
      <arg name="align_depth"              value="$(arg align_depth)"/>

      <arg name="fisheye_width"            value="$(arg fisheye_width)"/>
      <arg name="fisheye_height"           value="$(arg fisheye_height)"/>
      <arg name="enable_fisheye"           value="$(arg enable_fisheye)"/>

      <arg name="depth_width"              value="$(arg depth_width)"/>
      <arg name="depth_height"             value="$(arg depth_height)"/>
      <arg name="enable_depth"             value="true"/>

      <arg name="color_width"              value="$(arg color_width)"/>
      <arg name="color_height"             value="$(arg color_height)"/>
      <arg name="enable_color"             value="true"/>

      <arg name="infra_width"              value="$(arg infra_width)"/>
      <arg name="infra_height"             value="$(arg infra_height)"/>
      <arg name="enable_infra1"            value="false"/>
      <arg name="enable_infra2"            value="false"/>

      <arg name="fisheye_fps"              value="$(arg fisheye_fps)"/>
      <arg name="depth_fps"                value="6"/>
      <arg name="infra_fps"                value="$(arg infra_fps)"/>
      <arg name="color_fps"                value="6"/>
      <arg name="gyro_fps"                 value="$(arg gyro_fps)"/>
      <arg name="accel_fps"                value="$(arg accel_fps)"/>
      <arg name="enable_gyro"              value="$(arg enable_gyro)"/>
      <arg name="enable_accel"             value="$(arg enable_accel)"/>
      <arg name="filters"                  value="$(arg filters)"/>
    </include>

    <include if="$(arg rgb_processing)"
             file="$(find rgbd_launch)/launch/includes/rgb.launch.xml">
      <arg name="manager"                       value="$(arg manager)" />
      <arg name="respawn"                       value="$(arg respawn)" />
      <arg name="rgb"                           value="$(arg rgb)" />
      <arg name="debayer_processing"            value="$(arg debayer_processing)" />
    </include>

     <node pkg="nodelet" type="nodelet" name="register_depth"
            args="load depth_image_proc/register $(arg manager) $(arg bond)" respawn="$(arg respawn)">
        <remap from="rgb/camera_info"             to="$(arg rgb)/camera_info" />
        <remap from="depth/camera_info"           to="$(arg depth)/camera_info" />
        <remap from="depth/image_rect"            to="$(arg depth)/image_rect_raw" />
        <remap from="depth_registered/image_rect" to="$(arg depth_registered)/sw_registered/image_rect_raw" />
      </node>

      <!-- Publish registered XYZRGB point cloud with software registered input -->
      <node pkg="nodelet" type="nodelet" name="points_xyzrgb_sw_registered"
            args="load depth_image_proc/point_cloud_xyzrgb $(arg manager) $(arg bond)" respawn="$(arg respawn)">
        <remap from="rgb/image_rect_color"        to="$(arg rgb)/image_rect_color" />
        <remap from="rgb/camera_info"             to="$(arg rgb)/camera_info" />
        <remap from="depth_registered/image_rect" to="$(arg depth_registered_filtered)/sw_registered/image_rect_raw" />
        <remap from="depth_registered/points"     to="$(arg depth_registered)/points" />
      </node>

      <!-- Publish registered XYZRGB point cloud with hardware registered input (ROS Realsense depth alignment) -->
      <node pkg="nodelet" type="nodelet" name="points_xyzrgb_hw_registered"
            args="load depth_image_proc/point_cloud_xyzrgb $(arg manager) $(arg bond)" respawn="$(arg respawn)">
        <remap from="rgb/image_rect_color"        to="$(arg rgb)/image_rect_color" />
        <remap from="rgb/camera_info"             to="$(arg rgb)/camera_info" />
        <remap from="depth_registered/image_rect" to="$(arg depth_registered)/image_raw" />
        <remap from="depth_registered/points"     to="$(arg depth_registered_pub)/points" />
      </node>
<!--
      <node pkg="nodelet" type="nodelet"
        args="load image_proc/crop_decimate $(arg manager) $(arg bond)" respawn="$(arg respawn)"
        name="h3">
        <param name="x_offset" type="int" value="0" />
        <param name="y_offset" type="int" value="239" />
        <param name="width" type="int" value="640" />
        <param name="height" type="int" value="3" />

        <remap from="camera/image_raw" to="depth/image_rect_raw"/>
        <remap from="camera/camera_info" to="depth/camera_info"/>
  
        <remap from="camera_out/image_raw" to="depth/image_rect_raw_h3"/>
        <remap from="camera_out/camera_info" to="depth/camera_info_h3"/>
      </node>

      <node pkg="nodelet" type="nodelet"
        args="load image_proc/crop_decimate $(arg manager) $(arg bond)" respawn="$(arg respawn)"
        name="h1">
        <param name="decimation_y" type="int" value="1" />

        <remap from="camera/image_raw" to="depth/image_rect_raw_h3"/>
        <remap from="camera/camera_info" to="depth/camera_info_h3"/>
  
        <remap from="camera_out/image_raw" to="depth/image_rect_raw_h1"/>
        <remap from="camera_out/camera_info" to="depth/camera_info_h1"/>
      </node>

      <node pkg="nodelet" type="nodelet"
        args="load image_proc/crop_decimate $(arg manager) $(arg bond)" respawn="$(arg respawn)"
        name="v3">
        <param name="x_offset" type="int" value="319" />
        <param name="y_offset" type="int" value="0" />
        <param name="width" type="int" value="3" />
        <param name="height" type="int" value="480" />

        <remap from="camera/image_raw" to="depth/image_rect_raw"/>
        <remap from="camera/camera_info" to="depth/camera_info"/>
  
        <remap from="camera_out/image_raw" to="depth/image_rect_raw_v3"/>
        <remap from="camera_out/camera_info" to="depth/camera_info_v3"/>
      </node>

      <node pkg="nodelet" type="nodelet"
        args="load image_proc/crop_decimate $(arg manager) $(arg bond)" respawn="$(arg respawn)"
        name="v1">
        <param name="decimation_x" type="int" value="1" />

        <remap from="camera/image_raw" to="depth/image_rect_raw_v3"/>
        <remap from="camera/camera_info" to="depth/camera_info_v3"/>
  
        <remap from="camera_out/image_raw" to="depth/image_rect_raw_v1"/>
        <remap from="camera_out/camera_info" to="depth/camera_info_v1"/>
      </node>

-->

      <!-- <node pkg="nodelet" type="nodelet" output="screen"
        args="load pointcloud_to_laserscan/pointcloud_to_laserscan_nodelet $(arg manager) $(arg bond)" respawn="$(arg respawn)"
        name="pcl2laser">

        <param name="target_frame" type="str" value="camera_link" />
        <param name="min_height" type="double" value="0.4" />
        <param name="max_height" type="double" value="0.6" />
        <param name="min_range" type="double" value="0" />
        <param name="max_range" type="double" value="0" />
        <param name="angle_min" type="double" value="-0.78539816339" />
        <param name="angle_max" type="double" value="0.78539816339" />
        <param name="angle_increment" type="double" value="0.01745329251" />
        <param name="use_inf" type="boolean" value="true" />

        <remap from="cloud_in" to="depth_registered/points"/>

        <remap from="scan" to="pcl2laser/laserscan"/>
      </node> -->

      <node pkg="nodelet" type="nodelet" output="screen"
        args="load image_proc/scan $(arg manager) $(arg bond)" respawn="$(arg respawn)"
        name="image_proc_scan">
        <param name="h_x_offset" type="int" value="30" />
        <param name="h_y_offset" type="int" value="289" />
        <param name="h_width" type="int" value="580" />
        <param name="h_height" type="int" value="3" />

        <param name="v_x_offset" type="int" value="209" />
        <param name="v_y_offset" type="int" value="30" />
        <param name="v_width" type="int" value="3" />
        <param name="v_height" type="int" value="420" />

        <remap from="camera/image_raw" to="depth/image_rect_raw"/>
        <remap from="camera/image_info" to="depth/camera_info"/>
  
        <remap from="camera_out/h_scans" to="depth/h_scans"/>
        <remap from="camera_out/v_scans" to="depth/v_scans"/>
        <remap from="camera_out/h_scan" to="depth/h_scan"/>
      </node>

  
      <node pkg="nodelet" type="nodelet" output="screen"
        args="load depthimage_to_laserscan/DepthImageToLaserScanNodelet $(arg manager) $(arg bond)" respawn="$(arg respawn)"
        name="depth2laser">
          <param name="scan_height" type="int" value="240"/>
          <param name="range_max" type="double" value="100.0"/>
          <param name="range_min" type="double" value="-100.0"/>
          <param name="output_frame_id" type="str" value="base_link"/>
          <remap from="image" to="depth/image_rect_raw" />
          <remap from="scan" to="depth/laserscan" />
      </node>

  </group>
</launch>
